### DISCUSSION TITLE
Why qpdf/qpdf-dev?

### DISCUSSION THREAD

**Discussion by tillkamppeter** [qpdf/qpdf-dev#2] [General]
We are a free software project with QPDF and therefore we should discuss everything project-related in public. Therefore I suggest to give up/remove qpdf/qpdf-dev and do all our discussions on qpdf/qpdf. Let us copy over the one which got started here (#1) to qpdf/qpdf.

The only thing which we have to discuss in private a security vulnerabilities. And for that we do not need anything more than enabling qpdf/qpdf to receive private security bug reports (if not already done so). For enabling security bug rports I have written a HOWTO:

https://openprinting.github.io/OpenPrinting-News-May-2023/#handling-reported-security-bugs-with-github

WDYT? Should we stop with qpdf/qpdf-dev?

**Comment by m-holger** [comment]
I think there are 3 questions here:

1. Why have qpdf/qpdf-dev?
2. Should qpdf/qpdf-dev be private? 
3. Should qpdf support private raising of security related issues?

**1. Why have qpdf/qpdf-dev?**

The purpose is to provide a platform for discussion that is primarily or exclusively of interest to the developers of qpdf. This in itself does not feel particularly controversial to me.

**3. Should qpdf support private raising of security related issues?**

This sounds sensible to me.

**2. Should qpdf/qpdf-dev be private?** 

Looking at the purpose given under 1, I agree that the discussion should be public and accessible to anyone who wants to contribute to qpdf or is for any other reason interested in its inner workings.

A complication arises from the fact that the repository is also seen as a means of sharing some documents that Jay currently maintains privately. I haven't had time to look at the HOWTO properly and don't know whether it addresses the issue. In any case, it is probably better to treat this as a separate issue and make the discussions public.

**Comment by jberkenbilt** [comment]
The issue is that I have a large collection of PDF files that were attached to bug reports or shared with me privately. Right now, there are 847 files, though 727 of them are from issue qpdf/qpdf#476. I have notes on which ones are private. Most aren't. What I'll do is organize these and push them into this repository. Some files are large, so I will use LFS with those. Basically this repo can be a public repo to hold onto things like that that shouldn't be part of the main repo.

I agree that discussions about development should be in the main repo.

**Comment by m-holger** [comment]
For continuity, extract from private emails:

On Tue, Jun 18, 2024, at 12:04 PM, M Holger wrote:

> I think there is a case for keeping user level discussion and developer level discussions separate. I agree that the discussions should be public, but that can be achieved by providing the info / link on qpdf where the discussions can be found.

**Comment by m-holger** [comment]
To provide the background for creating qpdf-dev, over the last  couple of years there have been a  large number of private email exchanges between Jay and me discussing qpdf. Other discussions have taken place within PRs. 

There are a number of problems with this, including the difficulty of keeping track of discussions, ensuring there is some continuity, and ensuring that topics discussed don't simply fall off the radar.

On a more fundamental level, there is the problem that the points discussed in private emails are obviously not accessible to potential contributors and other interested parties. To make things worse, the 'big picture' discussions are more likely to be by email  and therefore the information that is publicly available in PRs etc is missing context and therefore does not make much sense to others. For example, there are a number of PRs that deal with the twin thread of ensuring that the object cache only contains valid objects and preparing for support of incremental updates but that at best hint at the context.

That leaves the obvious question: why don't we simply discuss all of this using the qpdf discussions area?

To quote Spock

> The Needs of the Many Outweigh the Needs of the Few

The majority of visitors to qpdf discussions are users of the qpdf CLI or ordinary library users. Even for library users, they may not have a knowledge of C++ as they  may be using the library indirectly through pikepdf  or through the C API.

In my view, a discussions forum should allow their average user to determine easily whether the information they are looking for is available / is already being discussed. Adding discussions that are not easily accessible to average users in my view is counter-productive and makes it more likely that users simply give up and go away. Significant developer discussion should be announced / summarized on qpdf in terms that are meaningful to ordinary users and a link provided, but I don't think that the discussions themself should clutter up the qpdf discussions area.

**Comment by jberkenbilt** [comment]
I am not comfortable with fragmenting discussion across two repos. I'm already finding myself having to look in more places than I want to. I don't agree that the needs of the many outweigh the needs of the few here. I think the needs of the people who spend a lot of their volunteer time outweigh the needs of the casual passers by. Rather than having a separate repo, we should organize the discussion forums in the main repo so that end users can find what they want. My preference is that we do this as soon as we can and migrate what's here over to the main repo while that's still a manageable task.

**Reply by m-holger** [reply]
If I look at two discussions, https://github.com/qpdf/qpdf/discussions/785 and https://github.com/qpdf/qpdf/discussions/1104, I cannot help but notice
- only people specifically invited contributed
- two days after opening the discussions participants were down to the three usual suspects
- only the initial post of a discussion received any reactions from non-participants.

If I have a staff meeting and the discussion is monopolized by three members of one team, my reaction tends to be to say "lets take this off-line and discuss it after the meeting. Anybody who is interested is welcome to stay" and than try to find a way to represent the topic in a way that engages the rest of the staff.  https://github.com/qpdf/qpdf-dev/pull/8 is one attempt of engaging the rest of the meeting. I have no idea whether it is going to work, but I think it is worth a try.

My concern is that by discussing the 'how' and 'when' in the same meeting as the 'what', discussion of the 'what' will be stifled, At the same time, the discussion of the how will be insufficiently granular to be useful.

**Reply by jberkenbilt** [reply]
Okay, well, I've made my opinion clear, and I believe you have considered it, so I will defer to you on this one. Let's just always make sure we have bidrectional links so it's easy for people to find what they need.

**Reply by m-holger** [reply]
I agree, bidrectional links are definitely needed.
